fsw/src/bpnode_app.c

```diff
@@ -0,0 +1,261 @@
+ * All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain
+ * a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ************************************************************************/
+
+/**
+ * \file
+ *   This file contains the source code for the BPNode App.
+ */
+
+/*
+** Include Files:
+*/
+#include "bpnode_app.h"
+#include "bpnode_cmds.h"
+#include "bpnode_utils.h"
+#include "bpnode_eventids.h"
+#include "bpnode_dispatch.h"
+#include "bpnode_tbl.h"
+#include "bpnode_version.h"
+
+/*
+** Global data
+*/
+BPNode_AppData_t BPNode_AppData;
+
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *  * *  * * * * **/
+/*                                                                            */
+/* Application entry point and main process loop                              */
+/*                                                                            */
+/* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *  * *  * * * * **/
+void BPNode_AppMain(void)
+{
+    CFE_Status_t     Status;
+    CFE_SB_Buffer_t *BufPtr;
+
+    /*
+    ** Create the first Performance Log entry
+    */
+    CFE_ES_PerfLogEntry(BPNODE_PERF_ID);
+
+    /*
+    ** Perform application-specific initialization
+    ** If the Initialization fails, set the RunStatus to
+    ** CFE_ES_RunStatus_APP_ERROR and the App will not enter the RunLoop
+    */
+    Status = BPNode_AppInit();
+    if (Status != CFE_SUCCESS)
+    {
+        BPNode_AppData.RunStatus = CFE_ES_RunStatus_APP_ERROR;
+    }
+
+    /*
+    ** BPNode Runloop
+    */
+    while (CFE_ES_RunLoop(&BPNode_AppData.RunStatus) == true)
+    {
+        /*
+        ** Performance Log Exit Stamp
+        */
+        CFE_ES_PerfLogExit(BPNODE_PERF_ID);
+
+        /* Pend on receipt of wakeup message */
+        Status = CFE_SB_ReceiveBuffer(&BufPtr, BPNode_AppData.WakeupPipe, 
+                                                BPNODE_WAKEUP_PIPE_TIMEOUT);
+
+        /*
+        ** Performance Log Entry Stamp
+        */
+        CFE_ES_PerfLogEntry(BPNODE_PERF_ID);
+
+        if (Status == CFE_SUCCESS)
+        {
+            /* Process wakeup tasks */
+            Status = BPNode_ProcessMain();
+        }
+        
+        if (Status != CFE_SUCCESS)
+        {
+            CFE_EVS_SendEvent(BPNODE_PIPE_ERR_EID, CFE_EVS_EventType_ERROR,
+                              "SB Pipe Read Error, App Will Exit");
+
+            BPNode_AppData.RunStatus = CFE_ES_RunStatus_APP_ERROR;
+        }
```

Comment:
This `CFE_SB_ReceiveBuffer ` can return `CFE_SB_TIME_OUT`, and we wouldn't want to exit if we get that.



@dzbaker  - would you prefer we keep track of SB timeouts, or silently ignore them, or something else?

Reply:
I can add that exception, yeah

Reply:
Perhaps we log a debug event for timeouts?

Reply:
How useful would that debug event be? I kinda like how HS does its flow (https://github.com/nasa/HS/blob/main/fsw/src/hs_app.c#L148), since it's got a similar wakeup pipe/ command pipe setup to ours

Reply:
I think following the HS flow makes the most sense.

Reply:
Sweet, I went ahead and pushed that already, lmk if it looks good to you

Reply:
I think so, but I guess if for some unexpected reason CFE_SB_NO_MESSAGE were returned, it shouldn't be treated like an error. I assume that's the rationale of the HS app developers at least, since that's the template I was following.